fvm/src/call_manager/default.rs

```diff
@@ -471,6 +477,153 @@ where
         })
     }
 
+    fn tst_validate(&mut self, msg: Message, sig: Signature) -> Result<()> {
+        // Lookup the actor.
+        let state = self
+            .state_tree()
+            .get_actor_id(msg.from.id().unwrap())?
+            .ok_or_else(|| syscall_error!(NotFound; "actor does not exist: {}", msg.from))?;
+
+        // Builtin or external?
+        // TODO
+
+        // Charge the method gas.
+        // TODO
+
+        // Store the parametrs, and initialize the block registry for the target actor.
+        let mut block_registry = BlockRegistry::new();
+        let params_id = block_registry.put(params)?;
+
+        // Increment invocation count
+        self.invocation_count += 1;
+
+        // This is a cheap operation as it doesn't actually clone the struct,
+        // it returns a referenced copy.
+        let engine = self.engine().clone();
+
+        log::trace!("validating message {} -> {}::{}", msg.to, msg.from, sig.sig_type);
+        self.map_mut(|cm| {
```

Comment:
I assume there's a bunch of logic we can deduplicate here? Do you want to venture and do it?

Reply:
there is. I don't want to say yes just yet since while they do have lots of similar code, their flows could end up being different (including the upgrade call in this as well)